File: backend/utils/stateManager.js
// ─── State Definitions ─────────────────────────────────────────────────────────
const STATES = {
  JOB_SELECTION:      'JOB_SELECTION',
  LOCATION_SELECTION: 'LOCATION_SELECTION',
  SHIFT_SELECTION:    'SHIFT_SELECTION',
  COLLECT_NAME:       'COLLECT_NAME',
  COLLECT_EMAIL:      'COLLECT_EMAIL',
  COLLECT_PHONE:      'COLLECT_PHONE',
  COLLECT_DOB:        'COLLECT_DOB',
  COLLECT_START_DATE: 'COLLECT_START_DATE',
  COLLECT_RESUME:     'COLLECT_RESUME',
  COLLECT_LINKEDIN:   'COLLECT_LINKEDIN',
  CONFIRMATION:       'CONFIRMATION',
  SUBMITTED:          'SUBMITTED',
};

// ─── Job Data ──────────────────────────────────────────────────────────────────
const JOBS = [
  'Forklift Operator',
  'General Labour',
  'Skidders & Packers',
  'Mechanical Helper',
  'Machine Operators',
  'Light Loading/Unloading',
  'Sanitation Workers',
  'Chemical Compounder',
];

const LOCATIONS = ['Etobicoke', 'Tottenham', 'Mississauga'];

const SHIFTS = ['Mornings', 'Afternoons', 'Nights', 'Flexible'];

// ─── Ordered Flow ──────────────────────────────────────────────────────────────
// This is the canonical sequence — the state machine follows this exactly
const STATE_FLOW = [
  STATES.JOB_SELECTION,
  STATES.LOCATION_SELECTION,
  STATES.SHIFT_SELECTION,
  STATES.COLLECT_NAME,
  STATES.COLLECT_EMAIL,
  STATES.COLLECT_PHONE,
  STATES.COLLECT_DOB,
  STATES.COLLECT_START_DATE,
  STATES.COLLECT_RESUME,
  STATES.COLLECT_LINKEDIN,
  STATES.CONFIRMATION,
  STATES.SUBMITTED,
];

// Maps a field name (from edit requests) back to the state that collects it
const FIELD_TO_STATE = {
  job:        STATES.JOB_SELECTION,
  position:   STATES.JOB_SELECTION,
  role:       STATES.JOB_SELECTION,
  location:   STATES.LOCATION_SELECTION,
  shift:      STATES.SHIFT_SELECTION,
  name:       STATES.COLLECT_NAME,
  email:      STATES.COLLECT_EMAIL,
  phone:      STATES.COLLECT_PHONE,
  dob:        STATES.COLLECT_DOB,
  birthday:   STATES.COLLECT_DOB,
  birth:      STATES.COLLECT_DOB,
  start:      STATES.COLLECT_START_DATE,
  date:       STATES.COLLECT_START_DATE,
  resume:     STATES.COLLECT_RESUME,
  cv:         STATES.COLLECT_RESUME,
  linkedin:   STATES.COLLECT_LINKEDIN,
};

// ─── Helpers ───────────────────────────────────────────────────────────────────
const getNextState = (currentState) => {
  const idx = STATE_FLOW.indexOf(currentState);
  if (idx === -1 || idx >= STATE_FLOW.length - 1) return STATES.SUBMITTED;
  return STATE_FLOW[idx + 1];
};

// Quick-reply buttons to display per state
const getButtons = (state) => {
  switch (state) {
    case STATES.JOB_SELECTION:
      return JOBS;
    case STATES.LOCATION_SELECTION:
      return LOCATIONS;
    case STATES.SHIFT_SELECTION:
      return SHIFTS;
    case STATES.COLLECT_RESUME:
      return ["Skip — I'll send later"];
    case STATES.COLLECT_LINKEDIN:
      return ['Skip — No LinkedIn'];
    case STATES.CONFIRMATION:
      return ['Submit Application', 'Edit Details'];
    default:
      return [];
  }
};

// Input type hint so the widget knows what UI to render
const getInputType = (state) => {
  if (state === STATES.COLLECT_RESUME) return 'file';
  return 'text';
};

// The session data key that each state writes into
const getDataKey = (state) => {
  const map = {
    [STATES.JOB_SELECTION]:      'jobRole',
    [STATES.LOCATION_SELECTION]: 'location',
    [STATES.SHIFT_SELECTION]:    'shift',
    [STATES.COLLECT_NAME]:       'name',
    [STATES.COLLECT_EMAIL]:      'email',
    [STATES.COLLECT_PHONE]:      'phone',
    [STATES.COLLECT_DOB]:        'dob',
    [STATES.COLLECT_START_DATE]: 'startDate',
    [STATES.COLLECT_LINKEDIN]:   'linkedIn',
  };
  return map[state] || null;
};

// Detect if a user's edit message maps to a field jump
const detectEditField = (message) => {
  const lower = message.toLowerCase();
  for (const [keyword, state] of Object.entries(FIELD_TO_STATE)) {
    if (lower.includes(keyword)) return state;
  }
  return null;
};

module.exports = {
  STATES,
  JOBS,
  LOCATIONS,
  SHIFTS,
  STATE_FLOW,
  getNextState,
  getButtons,
  getInputType,
  getDataKey,
  detectEditField,
};

File: backend_py/utils/state_manager.py
from enum import Enum
from typing import Optional, List


# ─── States ───────────────────────────────────────────────────────────────────
class States(str, Enum):
    JOB_SELECTION      = "JOB_SELECTION"
    LOCATION_SELECTION = "LOCATION_SELECTION"
    SHIFT_SELECTION    = "SHIFT_SELECTION"
    COLLECT_NAME       = "COLLECT_NAME"
    COLLECT_EMAIL      = "COLLECT_EMAIL"
    COLLECT_PHONE      = "COLLECT_PHONE"
    COLLECT_DOB        = "COLLECT_DOB"
    COLLECT_START_DATE = "COLLECT_START_DATE"
    COLLECT_RESUME     = "COLLECT_RESUME"
    COLLECT_LINKEDIN   = "COLLECT_LINKEDIN"
    CONFIRMATION       = "CONFIRMATION"
    SUBMITTED          = "SUBMITTED"


# ─── Job data ─────────────────────────────────────────────────────────────────
JOBS = [
    "Forklift Operator",
    "General Labour",
    "Skidders & Packers",
    "Mechanical Helper",
    "Machine Operators",
    "Light Loading/Unloading",
    "Sanitation Workers",
    "Chemical Compounder",
]

LOCATIONS = ["Etobicoke", "Tottenham", "Mississauga"]

SHIFTS = ["Morning (6AM–2PM)", "Afternoon (2PM–10PM)", "Night (10PM–6AM)"]


# ─── Flow order ───────────────────────────────────────────────────────────────
STATE_FLOW = [
    States.JOB_SELECTION,
    States.LOCATION_SELECTION,
    States.SHIFT_SELECTION,
    States.COLLECT_NAME,
    States.COLLECT_EMAIL,
    States.COLLECT_PHONE,
    States.COLLECT_DOB,
    States.COLLECT_START_DATE,
    States.COLLECT_RESUME,
    States.COLLECT_LINKEDIN,
    States.CONFIRMATION,
    States.SUBMITTED,
]

# ─── Buttons per state ────────────────────────────────────────────────────────
STATE_BUTTONS = {
    States.JOB_SELECTION:      JOBS,
    States.LOCATION_SELECTION: LOCATIONS,
    States.SHIFT_SELECTION:    SHIFTS,
    States.COLLECT_RESUME:     ["Skip"],
    States.COLLECT_LINKEDIN:   ["Skip"],
    States.CONFIRMATION:       ["Confirm & Submit", "Edit"],
}

# ─── Input type per state ─────────────────────────────────────────────────────
STATE_INPUT_TYPE = {
    States.JOB_SELECTION:      "buttons",
    States.LOCATION_SELECTION: "buttons",
    States.SHIFT_SELECTION:    "buttons",
    States.COLLECT_RESUME:     "file",
    States.CONFIRMATION:       "buttons",
}

# ─── Which data key each state fills ─────────────────────────────────────────
STATE_DATA_KEY = {
    States.JOB_SELECTION:      "jobRole",
    States.LOCATION_SELECTION: "location",
    States.SHIFT_SELECTION:    "shift",
    States.COLLECT_NAME:       "name",
    States.COLLECT_EMAIL:      "email",
    States.COLLECT_PHONE:      "phone",
    States.COLLECT_DOB:        "dob",
    States.COLLECT_START_DATE: "startDate",
    States.COLLECT_RESUME:     "resumeFileName",
    States.COLLECT_LINKEDIN:   "linkedIn",
}

# ─── Edit keyword → field mapping ─────────────────────────────────────────────
FIELD_TO_STATE = {
    "jobRole":   States.JOB_SELECTION,
    "location":  States.LOCATION_SELECTION,
    "shift":     States.SHIFT_SELECTION,
    "name":      States.COLLECT_NAME,
    "email":     States.COLLECT_EMAIL,
    "phone":     States.COLLECT_PHONE,
    "dob":       States.COLLECT_DOB,
    "startDate": States.COLLECT_START_DATE,
    "resume":    States.COLLECT_RESUME,
    "linkedIn":  States.COLLECT_LINKEDIN,
}

EDIT_KEYWORDS = {
    "job": "jobRole", "role": "jobRole", "position": "jobRole",
    "location": "location", "city": "location",
    "shift": "shift", "schedule": "shift",
    "name": "name",
    "email": "email",
    "phone": "phone", "number": "phone",
    "dob": "dob", "birth": "dob", "date of birth": "dob",
    "start": "startDate", "available": "startDate",
    "resume": "resume", "cv": "resume",
    "linkedin": "linkedIn",
}


# ─── Helpers ──────────────────────────────────────────────────────────────────
def get_next_state(current: States) -> Optional[States]:
    try:
        idx = STATE_FLOW.index(current)
        if idx + 1 < len(STATE_FLOW):
            return STATE_FLOW[idx + 1]
    except ValueError:
        pass
    return None


def get_buttons(state: States) -> List[str]:
    return STATE_BUTTONS.get(state, [])


def get_input_type(state: States) -> str:
    return STATE_INPUT_TYPE.get(state, "text")


def get_data_key(state: States) -> Optional[str]:
    return STATE_DATA_KEY.get(state)


def detect_edit_field(message: str) -> Optional[str]:
    msg = message.lower()
    for keyword, field in EDIT_KEYWORDS.items():
        if keyword in msg:
            return field
    return None
